### PR TITLE
Fix for bootstrapping on NixOS

### PR DESCRIPTION
**Pull request by canndrew**
NixOS puts Linux's dynamic loader in wierd place. Detect when we're on NixOS and patch the downloaded bootstrap executables appropriately.

### CODE REVIEW

File: src/bootstrap/bootstrap.py [hunk 1]
```diff
@@ -185,9 +186,63 @@ def download_stage0(self):
...
+            self.fix_executable(self.bin_root() + "/bin/cargo")
             with open(self.cargo_stamp(), 'w') as f:
                 f.write(self.stage0_cargo_rev())
 
+    def fix_executable(self, fname):
+        # If we're on NixOS we need to change the path to the dynamic loader
+
+        default_encoding = sys.getdefaultencoding()
+        try:
+            ostype = subprocess.check_output(['uname', '-s']).strip().decode(default_encoding)
```
Comment (xen0n) [inline src/bootstrap/bootstrap.py]: Use `platform.system()` instead (with appropriate import added to the top). `uname` is not present on Windows and this gets executed before we know the OS type.

File: src/bootstrap/bootstrap.py [hunk 2]
```diff
@@ -89,7 +89,6 @@ def verify(path, sha_path, verbose):
                "    expected: {}".format(found, expected))
     return verified
 
-
```
Comment (xen0n) [inline src/bootstrap/bootstrap.py]: Style cleanup is best done in a separate commit.

File: src/bootstrap/bootstrap.py [hunk 3]
```diff
@@ -185,9 +186,63 @@ def download_stage0(self):
...
+        if ostype != "Linux":
+            return
+
+        if not os.path.exists("/nix/store"):
+            return
+        if os.path.exists("/lib"):
+            return
+
+        # At this point we're pretty sure the user is running NixOS
+        print("Info: you seem to be running NixOS. Attempting to patch " + fname)
```
Comment (xen0n) [inline src/bootstrap/bootstrap.py]: FWIW we're using all-lowercase messages in rustbuild, and this `Info` should be lower-cased too.

File: src/bootstrap/bootstrap.py [hunk 4]
```diff
@@ -185,9 +186,63 @@ def download_stage0(self):
...
+
+        try:
+            interpreter = subprocess.check_output(["patchelf", "--print-interpreter", fname])
+            interpreter = interpreter.strip().decode(default_encoding)
+        except subprocess.CalledProcessError as e:
+            print("Warning: failed to call patchelf: %s" % e)
```
Comment (xen0n) [inline src/bootstrap/bootstrap.py]: So should this `Warning` and some others below.

**Review by xen0n — CHANGES_REQUESTED** [review]
LGTM from a Python programmer's perspective, although I can't test this. Take care of the portability and this should be good to merge!

**Comment by alexcrichton** [comment]
@canndrew could you describe why this is necessary, and why we should do it as part of the build system? This seems like a lot of brittle logic easily broken for something that I'd naively consider to be a local problem, but I'm sure I'm not aware of all the details!

**Comment by brson** [comment]
I think this is a reasonable workaround, but it's a little weird - if you happen to be running on nix, it changes what it does to the outputs. In other words, one still can't build for NixOS generally, but if you happen to be building on Nix you get a Nix-compatible binary.

File: src/bootstrap/bootstrap.py (hunk 1 above)
Comment (canndrew) [inline src/bootstrap/bootstrap.py]: I was aware of `platform.system()` but I copied what was done [here](https://github.com/rust-lang/rust/blob/10f6a5c4431e09d355a0ba319a630e02a1e38361/src/bootstrap/bootstrap.py#L316) because I figured there must be a reason for it. This shouldn't break on windows because the `fix_executable` function just returns without doing anything if something goes wrong.

Comment (xen0n) [inline src/bootstrap/bootstrap.py]: Well, yeah. Didn't see that `WindowsError` right below. Although IMO the `uname` calls are only there for consistency with the `configure` logic.

**Comment by canndrew** [comment]
@alexcrichton 
The stage0 linux rust binaries that the build system downloads are hard-coded to look for the dynamic loader at `/lib/ld-linux-$arch.so.2`. NixOS is different to other linuxes and keeps its dynamic loader(s) under `/nix/store`, but it comes with a tool called patchelf for patching precompiled binaries to use a different dynamic loader path. The brittle logic shouldn't be a problem because the `fix_executable` function will just return without doing anything if it fails.

@brson 
Yeah, this isn't really the most correct way to do this on NixOS. Ideally you'd write a Nix derivation which takes the dynamic loader as an input so it can find the path to patch. Just using the dynamic loader for `sh` is very much a hack. However I want to just be able to `cd` into a rust directory and build it without screwing around with the nix tools. This will let me do that. And it shouldn't interfere with any derivations people have written if it patches the path immediately after downloading the binary - the derivation can always just patch it again with another path.

**Comment by alexcrichton** [comment]
Thanks for the explanation @canndrew!

**Comment by canndrew** [comment]
It might be a good idea to see what other NixOS users think of this. Maybe @Ericson2314 has an opinion? I'm not sure who else uses NixOS.

**Comment by Ericson2314** [comment]
Granted, I'm a little out of date on this stuff as I mainly just patch the pre-compiled binaries, and then that stopped working great because the Cargo ones moved. But I thought at least the official nixpkgs rustc package downloads the bootstrap binaries on rustbuild's behalf?

In any event, I think this stuff can go away once the cargo integration with out build systems + distributed cache stuff is figured out, and stdlib-aware Cargo is implemented. Then we can build rustc like any other binary very nicely. So hopefully this would be a temporary thing for less than the year 🤞.

File: src/bootstrap/bootstrap.py [hunk 5]
```diff
@@ -185,9 +186,63 @@ def download_stage0(self):
...
+        default_encoding = sys.getdefaultencoding()
+        try:
+            ostype = subprocess.check_output(['uname', '-s']).strip().decode(default_encoding)
+        except (subprocess.CalledProcessError, WindowsError):
+            return
+
+        if ostype != "Linux":
+            return
+
+        if not os.path.exists("/nix/store"):
```
Comment (nathan7) [inline src/bootstrap/bootstrap.py]: Any NixOS system has `/etc/NIXOS` (`nixos-rebuild` / the activation scripts check for its existence and refuse to run without it), so that's probably a better test.

Comment (canndrew) [inline src/bootstrap/bootstrap.py]: Good point. Thanks.

Comment (8573) [inline src/bootstrap/bootstrap.py]: nathan7's recommendation was for `/etc/NIXOS`, and I believe one could have a NixOS installation without `/etc/nixos`.

Comment (canndrew) [inline src/bootstrap/bootstrap.py]: Oh, I was wondering why he wrote it in uppercase. I didn't realize that file existed.

**Comment by nagisa** [comment]
Instead of patching the rustc binary itself, change bootstrap logic to invoke rustc with `ld.so bin/rustc`. This can be done inside the rustc wrapper for least churn (and similar wrapper can be made for cargo as well).

**Comment by nathan7** [comment]
@canndrew `/etc/nixos` is not necessarily present, especially in containers etc. `/etc/nixos` != `/etc/NIXOS`

File: src/bootstrap/bootstrap.py (hunk 5 above)
Comment (nathan7) [inline src/bootstrap/bootstrap.py]: for the record, `s/he/they/`

**Comment by xen0n** [comment]
@nagisa If this is implemented we'd want to change rustup as well, as one can `rustup toolchain link` the stage0 for whatever reason and expect things to work. Personally I just feel it's adding inconsistency as the places that end up being touched for the `ld.so` invocations are more than if only the binaries are patched and everything goes as usual.

**Comment by canndrew** [comment]
@nagisa What's the advantage of doing it like that? From my point of view, the precompiled rustc binaries are just broken for NixOS - they point to a file that doesn't exist. So why not just fix the problem directly?

**Comment by solson** [comment]
This is also the standard way to fix binaries on NixOS.

File: src/bootstrap/bootstrap.py (hunk 1 above)
Comment (alexcrichton) [inline src/bootstrap/bootstrap.py]: Could this deduplicate between here and the other location?

File: src/bootstrap/bootstrap.py (hunk 3 above)
Comment (alexcrichton) [inline src/bootstrap/bootstrap.py]: (agreed)
